### ISSUE TITLE
Look into derived typeclass instances for Coin

### ISSUE THREAD

**Issue by erikd** [IntersectMBO/cardano-ledger#1475]
Currently `Coin` is defined as:
```
newtype Coin = Coin Integer
  deriving (Show, Eq, Ord, Num, Integral, Real, Enum, NoUnexpectedThunks, Generic)
```

Questions a reasonable person could have with regard this type as it is now:

* Why does it have `Real` and what happens with float rounding?
* Why does it have `Num` and what happens when `+` or `*` overflows `maxLovelace` or `-` goes negative?
* Why are we newtype wrapping `Integer` and not `Word64` ?
* Why do we not provide functions for addition/subtraction etc which check for over/underflow?

If some of these typeclass instances are need for tests, then they should be defined as orphan instances in the tests. At the very least we should have comments addressing some of these points.

**Comment by dcoutts** [comment]
The Byron ledger did use a type that enforced the range, and this made all the operational partial which then needed checking all over the place, including in places where the overflows were impossible.

We concluded that it's a better design in the ledger to use an unbounded _signed_ internal representation. This means none of the operations are partial (no overflows), and we only have to have a single check in the UTxO tx validation, rather than dozens (on all the intermediate calculations).

We nearly retrofitted the Byron ledger along these lines, in this PR, but we decided to go all the way to `Integer` rather than `Natural`, and then the PR has bit-rotted. We will probably resurrect it once we get some time to sort these things out.

https://github.com/input-output-hk/cardano-ledger/pull/693

> At the very least we should have comments addressing some of these points.

Yes, we should document the rationale in the code.

**Comment by dcoutts** [comment]
The `Real` instance is indeed questionable.

**Comment by JaredCorduan** [comment]
the `Real` instance was needed for `Integral`, and `Integral` was needed for floor (when converting a`Rational` to a `Coin`.). I think it makes sense to remove these instances and provide a `rationalToCoinViaFloor` function.

**Comment by JaredCorduan** [comment]
closed by  #1609
